### PR TITLE
 Support warmup weight of provider.

### PR DESCRIPTION
**Pull request by NeGnail**
fix [issue#43](https://github.com/alipay/sofa-rpc/issues/43)

### CODE REVIEW

**Comment by khotyn** [comment]
Please change all the comment to English, and add a issue describe the bug in detail.

File: core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java [hunk 1]
```diff
@@ -71,28 +71,28 @@
...
+     * static properties, immutable.
      */
     private final ConcurrentHashMap<String, String>           staticAttrs      = new ConcurrentHashMap<String, String>();
 
     /**
-     * 动态属性，会动态变的 <br />
+     * dynamic properties change dynamically. <br />
      * <p>
-     * 例如动态权重，是否启用，预热标记等  invocationOptimizing
+     * For example dynamic weight, whether to enable, preheating mark, etc.  invocationOptimizing
```
Comment (ujjboy) [inline core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java]: The comment need more accurate.

File: core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java [hunk 2]
```diff
@@ -178,9 +178,22 @@ private ProviderInfo(final String url) {
...
+
+                            long warmupTime = CommonUtils.parseLong(kvpair[1], 0);
+                            this.setDynamicAttr(ProviderInfoAttrs.ATTR_WARMUP_TIME, warmupTime);
+
+                        } else if (ProviderInfoAttrs.ATTR_WARMUP_WEIGHT.equals(kvpair[0]) &&
+                            StringUtils.isNotEmpty(kvpair[1])) {
+
+                            int warmupWeight = CommonUtils.parseInt(kvpair[1], this.weight);
+                            this.setDynamicAttr(ProviderInfoAttrs.ATTR_WARMUP_WEIGHT, warmupWeight);
+
```
Comment (ujjboy) [inline core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java]: The parse operation of dynamic attributes move to registry helper.

File: core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java [hunk 3]
```diff
@@ -190,8 +203,10 @@ private ProviderInfo(final String url) {
                         } else {
                             this.staticAttrs.put(kvpair[0], kvpair[1]);
                         }
-
                     }
+
+                    processWarmUpWeight();
```
Comment (ujjboy) [inline core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java]: The parse operation of dynamic attributes move to registry helper.

File: core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java [hunk 4]
```diff
@@ -205,17 +220,38 @@ private ProviderInfo(final String url) {
     }
 
     /**
-     * 从thrift://10.12.120.121:9090 得到Provider
+     * Read the preheating weight parameter,
+     * decide whether to switch the state to the preheating period,
+     * and set the corresponding parameters during the preheating period.
+     */
+    private void processWarmUpWeight() {
```
Comment (ujjboy) [inline core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java]: The parse operation of dynamic attributes move to registry helper.

File: core/api/src/test/java/com/alipay/sofa/rpc/client/ProviderInfoTest.java [hunk 5]
```diff
@@ -37,6 +37,76 @@
...
+
+        Assert.assertEquals(now + "", providerInfo.getStaticAttr(ProviderInfoAttrs.ATTR_START_TIME));
+        Assert.assertEquals(300, providerInfo.getDynamicAttr(ProviderInfoAttrs.ATTR_WARMUP_WEIGHT));
+        Assert.assertEquals(5000L, providerInfo.getDynamicAttr(ProviderInfoAttrs.ATTR_WARMUP_TIME));
+        Assert.assertEquals(now + 5000, providerInfo.getDynamicAttr(ProviderInfoAttrs.ATTR_WARM_UP_END_TIME));
+
+        Assert.assertEquals(ProviderStatus.WARMING_UP, providerInfo.getStatus());
+        Assert.assertEquals(300, providerInfo.getWeight());
+
+        Thread.sleep(3000);
```
Comment (ujjboy) [inline core/api/src/test/java/com/alipay/sofa/rpc/client/ProviderInfoTest.java]: The sleep time is too long, change to 30 even more short.

File: test/test-integration/src/test/java/com/alipay/sofa/rpc/test/warmup/WarmUpTest.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        RpcRuntimeContext.destroy();
+        RpcInternalContext.removeContext();
+        RpcInvokeContext.removeContext();
+    }
+
+    @Test
+    public void testWarmUp() throws InterruptedException, NoSuchFieldException, IllegalAccessException {
+
+        RegistryConfig registryConfig = new RegistryConfig()
+            .setProtocol("zookeeper")
```
Comment (ujjboy) [inline test/test-integration/src/test/java/com/alipay/sofa/rpc/test/warmup/WarmUpTest.java]: If you want to run integration test with zookeeper, move test case to module `test-integration-3rd`.

File: extension-impl/registry-zk/src/main/java/com/alipay/sofa/rpc/registry/zk/ZookeeperRegistryHelper.java [hunk 7]
```diff
@@ -166,10 +167,7 @@ private static void addCommonAttrs(StringBuilder sb) {
         }
 
         for (ChildData childData : currentData) {
-            String url = childData.getPath().substring(providerPath.length() + 1); // 去掉头部
-            url = URLDecoder.decode(url, "UTF-8");
-            // byte[] data = childData.getData();
-            providerInfos.add(ProviderInfo.valueOf(url));
+            providerInfos.add(convertUrlToProvider(providerPath, childData));
```
Comment (ujjboy) [inline extension-impl/registry-zk/src/main/java/com/alipay/sofa/rpc/registry/zk/ZookeeperRegistryHelper.java]: The `URLDecoder.decode(url, "UTF-8")` should not be deleted.

File: extension-impl/registry-zk/src/main/java/com/alipay/sofa/rpc/registry/zk/ZookeeperRegistryHelper.java [hunk 8]
```diff
@@ -179,7 +177,44 @@ static ProviderInfo convertUrlToProvider(String providerPath, ChildData childDat
...
-        return ProviderInfo.valueOf(url);
+        ProviderInfo providerInfo = ProviderInfo.valueOf(url);
+
+        processWarmUpWeight(providerInfo);
+
+        return providerInfo;
+    }
+
+    /**
+     * Read the preheating weight parameter,
```
Comment (ujjboy) [inline extension-impl/registry-zk/src/main/java/com/alipay/sofa/rpc/registry/zk/ZookeeperRegistryHelper.java]: warmup

File: core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java [hunk 9]
```diff
@@ -461,6 +427,15 @@ public int getWeight() {
         return weight;
     }
 
+    /**
+     * Get the origin weight
+     *
+     * @return the weight
+     */
+    public int getOriginWeight() {
```
Comment (ujjboy) [inline core/api/src/main/java/com/alipay/sofa/rpc/client/ProviderInfo.java]: The weight is mutable, this method is no need here. If somewhere need to get origin weight of provider, just use `getStaticAttr("weight")`

File: extension-impl/registry-zk/src/main/java/com/alipay/sofa/rpc/registry/zk/ZookeeperRegistryHelper.java (hunk 7 above)
Comment (NeGnail) [reply extension-impl/registry-zk/src/main/java/com/alipay/sofa/rpc/registry/zk/ZookeeperRegistryHelper.java]: The method `convertUrlToProvider`  has the `URLDecoder.decode(url, "UTF-8")`.

**Pull request merged**